accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+Also, operating systems often offer opt-in or opt-out mechanisms to disallow runtime code generation, for example [Arbitrary Code Guard](https://docs.microsoft.com/en-us/windows/security/threat-protection/microsoft-defender-atp/customize-exploit-protection) (ACG) on Windows, [Allow Execution of JIT-compiled Code Entitlement](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-jit) on macOS.
+
+Mono and .NET Native for UWP have been shipping .NET runtimes that abide by the runtime code generation restriction. [Native AOT Form Factor](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT) experiment explores this space further.
+
+IL code interpreters circumvent the environment restriction on runtime code generation and allow .NET libraries that depend on runtime code generation to work, with lower performance. From a security point of view, IL code interpreters are not desirable in environments with disallowed code generation since IL code is equivalent of executable code as described above.
+
+## Out-of-process Jit-In-Time Compiler (JIT)
+
+Moving code generation into separate process has been implemented as a security mitigation in some cases, most recently in Chakra JavaScript engine. With this mitigation, the regular process does not have privilege to generate new code and it must delegate the code generation to dedicated process instead. The dedicated process with special privileges to introduce dynamically generated code is audited and heavily scrutinized. The advantage of this solution is that no code in the original process ever needs to write executable memory, so entire attack vectors (like compromised code altering memory mappings) can be shut down entirely. [Google Project Zero analyzed the Chakra implementation](https://github.com/googleprojectzero/p0tools/blob/master/JITServer/JIT-Server-whitepaper.pdf) and pointed out some of its weaknesses.
+
+We do not see strong need for introducing out of process JIT for .NET runtime in foreseeable future. The primary reason is that .NET does not support sand-boxed execution of untrusted code anymore. The out-of-proc JIT has advantages for execution of untrusted (verifiable) code. The security benefits of out-of-proc JIT for trusted unverifiable IL code are less interesting since strong soundness validation is not feasible for unverifiable IL code. The secondary reasons include availability of more secure options that disallow runtime code generation altogether, the implementation cost and necessity of a second process to run alongside the primary process.
+
```

Comment:
is it work noting here that shipping a compiler _at all_ has been a problem for restricted environments? Every now and again I still get requests on how to disable the .net compiler in framework installs. Or rosyln.